File: VScalc.m
function [F, VS_DetSig,VS_Sig,phi_DetSig,phi_Sig,sigSens,detSens,VS_DetSig2,VS_Sig2]=VScalc(f)

% This script uses ONHmodel.m and HBmodelEM.m to calculate vector strength
% at a given operating point
tic;
N = 50;    % Number of points to plot
Fmin = 0;   % Minimum force applied
Fmax = 200; % Maximum force applied
%f = 1/97;   % Frequency
F = Fmin:(Fmax-Fmin)/(N-1):Fmax;

ke = 100e-6;


for i = 1:N
    [Sig, DetSig, Time, Fext, Detf, Sigf] = ONHmodel(F(i),f);
    
    clear zs zd zf
    
    zs = smooth(Sig(500:end),length(Sig(500:end))/2);
    zd = smooth(DetSig(500:end),length(DetSig(500:end))/2);
    zf = smooth(Fext(500:end),length(Fext(500:end))/2);
    zdf = smooth(Detf(500:end),length(Detf(500:end))/2);
    zsf = smooth(Sigf(500:end),length(Sigf(500:end))/2);
    
    Sig(500:end) = Sig(500:end) - zs(1:length(Sig(500:end)))';
    DetSig(500:end) = DetSig(500:end) - zd(1:length(Sig(500:end)))';
    Fext(500:end) = Fext(500:end) - zf(1:length(Sig(500:end)))';
    Detf(500:end) = Detf(500:end) - zdf(1:length(Sig(500:end)))';
    Sigf(500:end) = Sigf(500:end) - zsf(1:length(Sig(500:end)))';
    
    Sighilb = hilbert(Sig(500:end));
    DetSighilb = hilbert(DetSig(500:end));
    Fexthilb = hilbert(Fext(500:end));
    Detfhilb = hilbert(Detf(500:end));
    Sigfhilb = hilbert(Sigf(500:end));
    
    Sigphi = angle(Sighilb);
    DetSigphi = angle(DetSighilb);
    Fextphi = angle(Fexthilb);
    Detfphi = angle(Detfhilb);
    Sigfphi = angle(Sigfhilb);
    
    
    phaseSig = Sigphi - Fextphi;
    phaseDetSig = DetSigphi - Fextphi;
    phaseSig2 = Sigphi - Detfphi;
    phaseDetSig2 = DetSigphi - Sigfphi;
    %phaseSig = Sigphi;
    %phaseDetSig = DetSigphi;    
    
    VS_Sig(i) = circ_r(phaseSig);
    VS_DetSig(i) = circ_r(phaseDetSig);
    VS_Sig2(i) = circ_r(phaseSig2);
    VS_DetSig2(i) = circ_r(phaseDetSig2);
    
    phi_Sig(i) = circ_mean(phaseSig);
    phi_DetSig(i) = circ_mean(phaseDetSig); 
    
    NFFT = numel(Time);
    f2 = (1000/2)*linspace(0,1,NFFT/2+1);

    actual_delta_f = 1000/numel(Time);
    %double sided
    detsigfft = fft(DetSig,NFFT)/numel(Time);
    
    q=find(f2>3 & f2<100);
    detAmp = max(abs(detsigfft(q)));
    detSens(i) = detAmp./(ke*F(i));


    actual_delta_f = 1000/numel(Time);
    %double sided
    sigfft = fft(Sig,NFFT)/numel(Time);
    
    q=find(f2>3 & f2<100);
    sigAmp = max(abs(sigfft(q)));
    sigSens(i) = sigAmp./(ke*F(i));
    %}
end

figure;
plot(F,VS_Sig,'r');hold on;plot(F,VS_DetSig,'k');
title('Vector Strength');
%{
figure;
plot(F,phi_Sig,'r');hold on;plot(F,phi_DetSig,'k');
title('Phase');
%}
figure;
plot(detSens,VS_DetSig,'k');hold on;plot(sigSens,VS_Sig,'r');
title('VS versus sensitivity');

figure;
plot(F,VS_Sig2,'r');hold on;plot(F,VS_DetSig2,'k');
title('Vector Strength 2')


toc;

File: VScalc.py

import numpy as np
import scipy
import matcompat

# if available import pylab (from matlibplot)
try:
    import matplotlib.pylab as plt
except ImportError:
    pass

def VScalc(f):

    # Local Variables: phaseSig, f2, Fextphi, zd, zf, detsigfft, phi_Sig, NFFT, Sighilb, DetSig, phaseDetSig2, VS_Sig2, zs, DetSigphi, VS_DetSig, VS_Sig, Fmax, Sigfhilb, DetSighilb, Sigfphi, detAmp, sigAmp, sigfft, Detfphi, Detfhilb, sigSens, VS_DetSig2, F, Sigphi, detSens, N, phi_DetSig, phaseSig2, actual_delta_f, Fmin, Time, Detf, Fexthilb, ke, Fext, i, f, Sigf, q, zsf, Sig, zdf, phaseDetSig
    # Function calls: plot, angle, figure, title, ONHmodel, max, fft, smooth, circ_mean, find, length, abs, circ_r, linspace, numel, VScalc, toc, hilbert, tic
    #% This script uses ONHmodel.m and HBmodelEM.m to calculate vector strength
    #% at a given operating point
    tic
    N = 50.
    #% Number of points to plot
    Fmin = 0.
    #% Minimum force applied
    Fmax = 200.
    #% Maximum force applied
    #%f = 1/97;   % Frequency
    F = np.arange(Fmin, (Fmax)+(matdiv(Fmax-Fmin, N-1.)), matdiv(Fmax-Fmin, N-1.))
    ke = 100e-6
    for i in np.arange(1., (N)+1):
        [Sig, DetSig, Time, Fext, Detf, Sigf] = ONHmodel(F[int(i)-1], f)
        clear(zszdzf)
        zs = smooth(Sig[499:], (length(Sig[499:])/2.))
        zd = smooth(DetSig[499:], (length(DetSig[499:])/2.))
        zf = smooth(Fext[499:], (length(Fext[499:])/2.))
        zdf = smooth(Detf[499:], (length(Detf[499:])/2.))
        zsf = smooth(Sigf[499:], (length(Sigf[499:])/2.))
        Sig[499:] = Sig[499:]-zs[0:length(Sig[499:])].conj().T
        DetSig[499:] = DetSig[499:]-zd[0:length(Sig[499:])].conj().T
        Fext[499:] = Fext[499:]-zf[0:length(Sig[499:])].conj().T
        Detf[499:] = Detf[499:]-zdf[0:length(Sig[499:])].conj().T
        Sigf[499:] = Sigf[499:]-zsf[0:length(Sig[499:])].conj().T
        Sighilb = hilbert(Sig[499:])
        DetSighilb = hilbert(DetSig[499:])
        Fexthilb = hilbert(Fext[499:])
        Detfhilb = hilbert(Detf[499:])
        Sigfhilb = hilbert(Sigf[499:])
        Sigphi = np.angle(Sighilb)
        DetSigphi = np.angle(DetSighilb)
        Fextphi = np.angle(Fexthilb)
        Detfphi = np.angle(Detfhilb)
        Sigfphi = np.angle(Sigfhilb)
        phaseSig = Sigphi-Fextphi
        phaseDetSig = DetSigphi-Fextphi
        phaseSig2 = Sigphi-Detfphi
        phaseDetSig2 = DetSigphi-Sigfphi
        #%phaseSig = Sigphi;
        #%phaseDetSig = DetSigphi;    
        VS_Sig[int(i)-1] = circ_r(phaseSig)
        VS_DetSig[int(i)-1] = circ_r(phaseDetSig)
        VS_Sig2[int(i)-1] = circ_r(phaseSig2)
        VS_DetSig2[int(i)-1] = circ_r(phaseDetSig2)
        phi_Sig[int(i)-1] = circ_mean(phaseSig)
        phi_DetSig[int(i)-1] = circ_mean(phaseDetSig)
        NFFT = numel(Time)
        f2 = np.dot(1000./2., np.linspace(0., 1., (NFFT/2.+1.)))
        actual_delta_f = 1000./numel(Time)
        #%double sided
        detsigfft = matdiv(np.fft(DetSig, NFFT), numel(Time))
        q = nonzero(np.logical_and(f2 > 3., f2<100.))
        detAmp = matcompat.max(np.abs(detsigfft[int(q)-1]))
        detSens[int(i)-1] = detAmp/np.dot(ke, F[int(i)-1])
        actual_delta_f = 1000./numel(Time)
        #%double sided
        sigfft = matdiv(np.fft(Sig, NFFT), numel(Time))
        q = nonzero(np.logical_and(f2 > 3., f2<100.))
        sigAmp = matcompat.max(np.abs(sigfft[int(q)-1]))
        sigSens[int(i)-1] = sigAmp/np.dot(ke, F[int(i)-1])
        #%}
        
    plt.figure
    plt.plot(F, VS_Sig, 'r')
    plt.hold(on)
    plt.plot(F, VS_DetSig, 'k')
    plt.title('Vector Strength')
    #%{
    plt.figure
    plt.plot(F, phi_Sig, 'r')
    plt.hold(on)
    plt.plot(F, phi_DetSig, 'k')
    plt.title('Phase')
    #%}
    plt.figure
    plt.plot(detSens, VS_DetSig, 'k')
    plt.hold(on)
    plt.plot(sigSens, VS_Sig, 'r')
    plt.title('VS versus sensitivity')
    plt.figure
    plt.plot(F, VS_Sig2, 'r')
    plt.hold(on)
    plt.plot(F, VS_DetSig2, 'k')
    plt.title('Vector Strength 2')
    toc
    return [F, VS_DetSig, VS_Sig, phi_DetSig, phi_Sig, sigSens, detSens, VS_DetSig2, VS_Sig2]